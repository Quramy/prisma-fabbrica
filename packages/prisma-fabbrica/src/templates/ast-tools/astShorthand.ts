import ts from "typescript";
import { pascalize } from "../../helpers";

type StripCreate<T extends string> = T extends `create${infer S}` ? Uncapitalize<S> : T;

export const ast = new Proxy(ts.factory, {
  get(target: any, n) {
    const name = n as string;
    if (name.startsWith("update")) return target[name];
    return target[`create${pascalize(name)}`];
  },
}) as unknown as {
  [K in keyof ts.NodeFactory as StripCreate<K>]: ts.NodeFactory[K];
};
