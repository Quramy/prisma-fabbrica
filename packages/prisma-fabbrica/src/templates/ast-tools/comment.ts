import ts from "typescript";

export function wrapWithTSDoc<T extends ts.Node>(text: string, node: T): T {
  const comment =
    "*\n" +
    text
      .trim()
      .split("\n")
      .map(line => ` * ${line}`)
      .join("\n") +
    "\n ";
  return ts.addSyntheticLeadingComment(node, ts.SyntaxKind.MultiLineCommentTrivia, comment, true);
}

export function insertLeadingBreakMarker<T extends ts.Node>(node: T): T {
  return ts.addSyntheticLeadingComment(node, ts.SyntaxKind.SingleLineCommentTrivia, "%BR%", true);
}
