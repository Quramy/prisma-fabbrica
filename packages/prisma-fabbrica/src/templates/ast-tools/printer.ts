import ts from "typescript";

export function createPrinter() {
  const printer = ts.createPrinter({
    removeComments: false,
    newLine: ts.NewLineKind.LineFeed,
    omitTrailingSemicolon: false,
  });

  return {
    print(sourceFile: ts.SourceFile) {
      const tmp = printer.printFile(sourceFile);
      return tmp.replace(/\/\/\s*%BR%/g, "");
    },
  };
}
