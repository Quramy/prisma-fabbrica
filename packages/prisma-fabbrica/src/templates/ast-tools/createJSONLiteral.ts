import ts from "typescript";
import { ast } from "./astShorthand";

type Primitive = string | boolean | number;

type JSONObjLike = {
  [key: string]: Primitive | JSONObjLike | JSONObjLike[];
};

type JSONLike = JSONObjLike | JSONObjLike[];

function createArrayLitreral(obj: ReadonlyArray<JSONObjLike>): ts.ArrayLiteralExpression {
  return ast.arrayLiteralExpression(obj.map(createObjectLiteral));
}

function createObjectLiteral(obj: JSONObjLike): ts.ObjectLiteralExpression {
  return ast.objectLiteralExpression(
    Object.entries(obj).map(([k, v]) =>
      ast.propertyAssignment(
        k,
        typeof v === "string"
          ? ast.stringLiteral(v)
          : typeof v === "number"
          ? ast.numericLiteral(v)
          : typeof v === "boolean"
          ? v
            ? ast.true()
            : ast.false()
          : Array.isArray(v)
          ? createArrayLitreral(v)
          : typeof v === "object"
          ? createObjectLiteral(v)
          : (null as never),
      ),
    ),
    true,
  );
}

export function createJSONLiteral(obj: JSONLike): ts.Expression {
  return Array.isArray(obj) ? createArrayLitreral(obj) : createObjectLiteral(obj);
}
