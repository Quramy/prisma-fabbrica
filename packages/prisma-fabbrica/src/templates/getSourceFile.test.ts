import { getDMMF } from "@prisma/internals";

import { getSourceFile } from ".";
import { createPrinter } from "./ast-tools/printer";

describe(getSourceFile, () => {
  it("generates TypeScript AST", async () => {
    const dmmf = await getDMMF({
      datamodel: `
        model User {
          id Int @id
          name String
        }
      `,
    });
    const sourceFile = getSourceFile({ document: dmmf });
    const printer = createPrinter();
    expect(printer.print(sourceFile)).toMatchSnapshot();
  });
});
