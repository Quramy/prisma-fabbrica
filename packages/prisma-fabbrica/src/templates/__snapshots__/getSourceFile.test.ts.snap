// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSourceFile generates TypeScript AST 1`] = `
"import { User } from "@prisma/client";
import { Prisma } from "@prisma/client";
import type { PrismaClient } from "@prisma/client";
import { getClient, ModelWithFields, createScreener, scalarFieldValueGenerator, Resolver, normalizeResolver, normalizeList, getSequenceCounter, } from "@quramy/prisma-fabbrica/lib/internal";
export { initialize, resetSequence } from "@quramy/prisma-fabbrica/lib/internal";
type BuildDataOptions = {
    readonly seq: number;
};
const modelFieldDefinitions: ModelWithFields[] = [{
        name: "User",
        fields: []
    }];
type UserScalarOrEnumFields = {
    id: number;
    name: string;
};
type UserFactoryDefineInput = {
    id?: number;
    name?: string;
};
type UserFactoryDefineOptions = {
    defaultData?: Resolver<UserFactoryDefineInput, BuildDataOptions>;
};
interface UserFactoryInterface {
    readonly _factoryFor: "User";
    build(inputData?: Partial<Prisma.UserCreateInput>): PromiseLike<Prisma.UserCreateInput>;
    buildCreateInput(inputData?: Partial<Prisma.UserCreateInput>): PromiseLike<Prisma.UserCreateInput>;
    buildList(inputData: number | readonly Partial<Prisma.UserCreateInput>[]): PromiseLike<Prisma.UserCreateInput[]>;
    pickForConnect(inputData: User): Pick<User, "id">;
    create(inputData?: Partial<Prisma.UserCreateInput>): PromiseLike<User>;
    createList(inputData: number | readonly Partial<Prisma.UserCreateInput>[]): PromiseLike<User[]>;
    createForConnect(inputData?: Partial<Prisma.UserCreateInput>): PromiseLike<Pick<User, "id">>;
}
function autoGenerateUserScalarsOrEnums({ seq }: {
    readonly seq: number;
}): UserScalarOrEnumFields {
    return {
        id: scalarFieldValueGenerator.Int({ modelName: "User", fieldName: "id", isId: true, isUnique: false, seq }),
        name: scalarFieldValueGenerator.String({ modelName: "User", fieldName: "name", isId: false, isUnique: false, seq })
    };
}
function defineUserFactoryInternal({ defaultData: defaultDataResolver }: UserFactoryDefineOptions): UserFactoryInterface {
    const seqKey = {};
    const getSeq = () => getSequenceCounter(seqKey);
    const screen = createScreener("User", modelFieldDefinitions);
    const build = async (inputData: Partial<Prisma.UserCreateInput> = {}) => {
        const seq = getSeq();
        const requiredScalarData = autoGenerateUserScalarsOrEnums({ seq });
        const resolveValue = normalizeResolver<UserFactoryDefineInput, BuildDataOptions>(defaultDataResolver ?? {});
        const defaultData = await resolveValue({ seq });
        const defaultAssociations = {};
        const data: Prisma.UserCreateInput = { ...requiredScalarData, ...defaultData, ...defaultAssociations, ...inputData };
        return data;
    };
    const buildList = (inputData: number | readonly Partial<Prisma.UserCreateInput>[]) => Promise.all(normalizeList(inputData).map(data => build(data)));
    const pickForConnect = (inputData: User) => ({
        id: inputData.id
    });
    const create = async (inputData: Partial<Prisma.UserCreateInput> = {}) => {
        const data = await build(inputData).then(screen);
        return await getClient<PrismaClient>().user.create({ data });
    };
    const createList = (inputData: number | readonly Partial<Prisma.UserCreateInput>[]) => Promise.all(normalizeList(inputData).map(data => create(data)));
    const createForConnect = (inputData: Partial<Prisma.UserCreateInput> = {}) => create(inputData).then(pickForConnect);
    return {
        _factoryFor: "User" as const,
        build,
        buildList,
        buildCreateInput: build,
        pickForConnect,
        create,
        createList,
        createForConnect,
    };
}
/**
 * Define factory for {@link User} model.
 *
 * @param options
 * @returns factory {@link UserFactoryInterface}
 */
export function defineUserFactory(options: UserFactoryDefineOptions = {}): UserFactoryInterface {
    return defineUserFactoryInternal(options);
}
"
`;
